### PR TITLE
Fixes for C++ exceptions

### PR DESCRIPTION
**Pull request by bsdjhb**
Generally this does not affect the layout of the first level of DWARF tables like LSDA and FDE, etc.  Instead, when exception info refers to a pointer, the toolchain now emits a relocation for a full capability for the stub symbols used to indirect to the personality function and C++ type info objects.  It then updates libunwind to assumes that pointers are capabilities and not offsets.

### CODE REVIEW

File: libunwind/src/AddressSpace.hpp [hunk 1]
```diff
@@ -578,6 +578,11 @@ inline bool LocalAddressSpace::findUnwindSections(pint_t targetAddr,
 #ifdef __CHERI_PURE_CAPABILITY__
         check_same_type<__uintcap_t, decltype(pinfo->dlpi_addr)>();
         check_same_type<const Elf_Phdr *, decltype(pinfo->dlpi_phdr)>();
+
+        // Cannot use CTestSubset here because the dpli_addr perms are a strict
+        // subset that never includes execute and so won't match targetAddr
+        // which is always executable.
+        //
```
Comment (nwf) [inline libunwind/src/AddressSpace.hpp]: You could use `__builtin_cheri_test_subset(__builtin_cheri_perms_and(cbdata->targetAddr, 0), pinfo->dlpi_addr)` to moot the permissions check?  But maybe that's too obfuscated. :)

File: libunwind/src/AddressSpace.hpp [hunk 2]
```diff
@@ -669,8 +669,32 @@ inline bool LocalAddressSpace::findUnwindSections(pint_t targetAddr,
...
         if (found_obj && found_hdr) {
           CHERI_DBG("found_obj && found_hdr in %s\n", pinfo->dlpi_name);
-          cbdata->sects->dwarf_section_length = object_length;
-          return true;
+
+	  // Find the PT_LOAD containing .eh_frame.
+          for (Elf_Half i = 0; i < pinfo->dlpi_phnum; i++) {
+            const Elf_Phdr *phdr = &pinfo->dlpi_phdr[i];
+            if (phdr->p_type == PT_LOAD) {
+              uintptr_t begin = getPhdrCapability(pinfo, phdr);
```
Comment (arichardson) [inline libunwind/src/AddressSpace.hpp]: I'm not sure my implementation of getPhdrCapability is quite right but if it works here LGTM.

File: libunwind/src/AddressSpace.hpp [hunk 3]
```diff
@@ -238,11 +238,6 @@ class _LIBUNWIND_HIDDEN LocalAddressSpace {
     return get<v128>(addr);
   }
   capability_t     getCapability(pint_t addr) { return get<capability_t>(addr); }
-  addr_t           getAddr(pint_t addr) {
-    // FIXME: for CHERI the actually address is the second 8 byte sequence
-    return get<addr_t>(addr);
-  }
-  __attribute__((always_inline))
   uintptr_t       getP(pint_t addr);
```
Comment (arichardson) [inline libunwind/src/AddressSpace.hpp]: Why did you remove the always_inline?

File: libunwind/src/AddressSpace.hpp (hunk 2 above)
Comment (bsdjhb) [reply libunwind/src/AddressSpace.hpp]: It works currently.  I might look at it it further.  It should probably be using assert_pointer_in_bounds() or some such not just checking the tag.  Hmm, I think more thought is required.

File: libunwind/src/AddressSpace.hpp (hunk 3 above)
Comment (bsdjhb) [reply libunwind/src/AddressSpace.hpp]: Misparsed this as belonging to get<addr_t>, I will put it back.
